docs/proposals/05-etcd-operator-tasks.md

```diff
@@ -0,0 +1,396 @@
+In this scenario, controller (say EtcdOperatorTasks Controller) will detect two pending tasks at the same time. The question that arises which task should be executed first ?
+
+To deal with such scenario, authors are proposing that each "out-of-bands" operator task can have some pre-defined pre-conditions which should be necessary to perform that operation otherwise controller can simply `reject` the operations.
+Moreover, controller (EtcdOperatorTasks Controller) will execute each operator task in a First-In-First-Out (FIFO) fashion.
+
+| Out-of-band Tasks                | Pre-conditions      |  Reasons                                               |
+| ---------------------------------|---------------------|--------------------------------------------------------|
+| `QuorumLossRecoveryTask`         | `replicas` > 1      | if no. of etcd replicas is one then it's simple restoration case, not a quorum loss recovery.         |
+| `OnDemandSnapshotTask`           | cluster should be healhty i.e quorum should be maintained        | taking a on-demand snapshot(full/delta) involves etcd cluster.           |
+| `OnDemandSnapCompactionTask`     | NA          | on-demand snapshot-compaction runs in a seprate pod, hence it don't depend on health of etcd cluster members or any other conditions.       |
+| `OnDemandMaintenanceTaskConfig`  | cluster should be healhty i.e quorum should be maintained         | on-demand etcd cluster maintenance depends on the health of etcd cluster.              |
+| `CopyBackupsTask`                | NA         | copy-backup tasks run seprately without affecting or required help from etcd cluster.                      |
+
+Steps to follow in case of multiple operator tasks:
+
+- Check for duplicate task. If found simply `reject` the task.
+- Verify if the pre-conditions of the task are met. If not, `reject` the task.
+   * If the pre-conditions are satisfied, add the task to the queue.
+
+- For adding any new operations to "out-of-band" operator tasks, user must have to define the pre-conditions of that task, if any.
+
+### Lifecycle of an Operator Tasks custom resources
+
+#### Creation
+
+* For [alpha feature](#alpha), creation of operator tasks custom resources can be done manually.
+* For [GA](#beta---ga), the creation of operator tasks custom resources will done by `druidctl add task <taskName>`.
+
+#### Deletion
+
+To garbage collect the custom resources can be done by another controller (say TTL controller) with help of `.spec.ttlSecondsAfterFinished` specified in each operator tasks custom resource.
+
+### Few Implementation Details
+
+1. In QuorumLossRecoveryTask, recovery controller can't move to next operation until pervious operation moved to `complete` state.
+2. If both type of maintenance is requested then controller should trigger etcd-compaction first then defragmentation.
```

Comment:
> OnDemandETCDCompactionTask waits on PermanentQuorumLossRecoveryTask

There is no real `wait-on` relationship here but rather a `pre-condition` behavior which fits a bit better. Pre-conditions for `OnDemandETCDCompactionTask` are:

1. There should not be an existing `in-progress` task for the same etcd
2. Quorum should be established
And maybe we come with more pre-conditions or acceptance criteria for even accepting this task. If even one of these conditions are not met we reject the task and record the rejection in its status.

Reply:
There are some tasks which can be run concurrently and some which perhaps needs to be sequenced. Need to revisit if this can easily be addressed by just defining pre-conditions.

Reply:
For this version of the proposal we only define `pre-condition` which serves as a guard to accept or reject the task. Everything by default will be queued. In the next increment we will enhance this to provide both queueing with ordering along with concurrent execution capabilities.